src/btree/bt_split.c

```diff
@@ -1509,30 +1516,59 @@ __split_multi_inmem(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi, WT
  *     Discard moved update lists from the original page.
  */
 static void
-__split_multi_inmem_final(WT_PAGE *orig, WT_MULTI *multi)
+__split_multi_inmem_final(WT_SESSION_IMPL *session, WT_PAGE *orig, WT_MULTI *multi)
 {
     WT_SAVE_UPD *supd;
+    WT_UPDATE *upd;
     uint32_t i, slot;
 
+    /* If we have saved updates, we must have decided to restore them to the new page. */
+    WT_ASSERT(session, multi->supd_entries == 0 || multi->supd_restore);
+
     /*
      * We successfully created new in-memory pages. For error-handling reasons, we've left the
      * update chains referenced by both the original and new pages. We're ready to discard the
      * original page, terminate the original page's reference to any update list we moved.
      */
-    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd)
+    for (i = 0, supd = multi->supd; i < multi->supd_entries; ++i, ++supd) {
+        /* Only need to restore update chain that has updates newer than the on page value. */
+        if (!supd->restore)
+            continue;
```

Comment:
I'm not sure why we're testing `supd.restore` here (and the comment talks about "restore")?

Reply:
For the individual update chain, we only need to restore it if it has newer updates than the onpage value written to the data file. Otherwise, we can discard it. Usually, not all the keys on a dirty page need to be restored. Here's the place we discard the keys that are clean after reconciliation.

Reply:
> Here's the place we discard the keys that are clean after reconciliation.

Right, that's why I'm thinking a comment that says "Only need to restore" doesn't help?

Reply:
Reword the comments.